README.md


[![pub package](https://img.shields.io/pub/v/haversine_distance.svg)](https://pub.dev/packages/haversine_distance)


# Haversine_Distance

A simple flutter package to calculate the distance between two coordinates

## Getting Started

The following project is a simple Dart package to calculate the distance between two points in latitude and longitude.

```dart
/// Import the package
import 'package:haversine_distance/haversine_distance.dart';

/// Create the two location objects you want to calculate the distance between.
/// The Location object is included in the package. 
final startCoordinate = new Location(60.389739, 5.322323);
final endCoordinate = new Location(60.393032, 5.327248);

/// Create a new haversine object
final haversineDistance = HaversineDistance();

/// Then calculate the distance between the two location objects and set a unit.
/// You can select between KM/MILES/METERS/NMI
haversineDistance.haversine(startCoordinate, endCoordinate, Unit.KM).floor();
 ```

 ## API

 ### Units to calculate
 KM, Meters, Miles, NMI,

 ```Dart
enum Unit { KM, MILE, METER, NMI }
 ```

 ### Location Class: 

```Dart
Class Location(){
 String latitude;
 String longitude;
}
```





## Thanks

The following Dart package is highly inspired by the following npm project: https://www.npmjs.com/package/haversine

CHANGELOG.md
## [1.2.1] Dart SDK

- Set the Dart lower SDK constraint to 2.12.0.

## [1.2.0] Null-safety

- Updated relevant packages to support Null Safety for Dart 2.0

## [1.1.0] - Added example file

- Added example project to improve pub.dev points. Pretty much the same as in test folder.

## [1.0.0] - Stable release

- First version tested in a real project

## [0.0.4] - Fixed Test

- Added tests to the project

## [0.0.1] - Initial release.

- The first version of the package. Not yet tested directly tested from pub.dev

test/haversine_test.dart
import 'package:haversine_distance/haversine_distance.dart';
import 'package:test/test.dart';

void main() {
  /// First coordinate test
  final startCoordinate = new Location(60.389739, 5.322323);
  final endCoordinate = new Location(60.393032, 5.327248);

  /// Second coordinate test
  final startCoordinate2 = new Location(60.382177, 5.334089);
  final endCoordinate2 = new Location(59.913040, 10.754140);

  test('Distance should be 455m', () {
    final haversineDistance = HaversineDistance();
    final distance = haversineDistance
        .haversine(startCoordinate, endCoordinate, Unit.METER)
        .floor();
    expect(distance, 455);
  });

  test('Distance should be 0.45m', () {
    final haversineDistance = HaversineDistance();
    final distance = haversineDistance
        .haversine(startCoordinate, endCoordinate, Unit.KM)
        .toStringAsFixed(2);
    expect(distance, '0.46');
  });

  test('Distance should be 455m', () {
    final haversineDistance = HaversineDistance();
    final distance = haversineDistance
        .haversine(startCoordinate, endCoordinate, Unit.METER)
        .floor();
    expect(distance, 455);
  });

  test('Bergen - Oslo in different formats', () {
    final haversineDistance = HaversineDistance();

    final distanceKm = haversineDistance
        .haversine(startCoordinate2, endCoordinate2, Unit.KM)
        .toStringAsFixed(2);

    final distanceMeter = haversineDistance
        .haversine(startCoordinate2, endCoordinate2, Unit.METER)
        .toStringAsFixed(2);

    final distanceNMI = haversineDistance
        .haversine(startCoordinate2, endCoordinate2, Unit.NMI)
        .toStringAsFixed(2);

    final distanceMile = haversineDistance
        .haversine(startCoordinate2, endCoordinate2, Unit.MILE)
        .toStringAsFixed(2);

    expect(distanceNMI, '164.36');

    expect(distanceMile, '189.21');

    expect(distanceMeter, '304405.91');

    expect(distanceKm, '304.41');
  });
}


lib/haversine.dart
library haversine;

import 'dart:math';

class Location {
  final double latitude;
  final double longitude;

  Location(this.latitude, this.longitude);
}

class RADII {
  int km;
  int mile;
  int meter;
  int nmi;

  RADII(this.km, this.mile, this.meter, this.nmi);
}

enum Unit { KM, MILE, METER, NMI }

class Haversine {
  final RADII radii = new RADII(6371, 3960, 6371000, 3440);

  double toRad(double num) {
    return num * pi / 180;
  }

  int getUnit(Unit unit) {
    switch (unit) {
      case (Unit.KM):
        return radii.km;
      case (Unit.MILE):
        return radii.mile;
      case (Unit.METER):
        return radii.meter;
      case (Unit.NMI):
        return radii.nmi;
      default:
        return radii.km;
    }
  }

  double haversine(
      Location startCoordinates, Location endCoordinates, Unit unit) {
    final R = getUnit(unit);
    final dLat = toRad(endCoordinates.latitude - startCoordinates.latitude);
    final dLon = toRad(endCoordinates.longitude - startCoordinates.longitude);
    final lat1 = toRad(startCoordinates.latitude);
    final lat2 = toRad(endCoordinates.latitude);

    final a = sin(dLat / 2) * sin(dLat / 2) +
        sin(dLon / 2) * sin(dLon / 2) * cos(lat1) * cos(lat2);
    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return R * c;
  }
}


lib/haversine_distance.dart
library haversine;

import 'dart:math';

class Location {
  final double latitude;
  final double longitude;

  Location(this.latitude, this.longitude);
}

class RADII {
  int km;
  int mile;
  int meter;
  int nmi;

  RADII(this.km, this.mile, this.meter, this.nmi);
}

enum Unit { KM, MILE, METER, NMI }

class HaversineDistance {
  final RADII radii = new RADII(6371, 3960, 6371000, 3440);

  double toRad(double num) {
    return num * pi / 180;
  }

  int getUnit(Unit unit) {
    switch (unit) {
      case (Unit.KM):
        return radii.km;
      case (Unit.MILE):
        return radii.mile;
      case (Unit.METER):
        return radii.meter;
      case (Unit.NMI):
        return radii.nmi;
      default:
        return radii.km;
    }
  }

  double haversine(
      Location startCoordinates, Location endCoordinates, Unit unit) {
    final R = getUnit(unit);
    final dLat = toRad(endCoordinates.latitude - startCoordinates.latitude);
    final dLon = toRad(endCoordinates.longitude - startCoordinates.longitude);
    final lat1 = toRad(startCoordinates.latitude);
    final lat2 = toRad(endCoordinates.latitude);

    final a = sin(dLat / 2) * sin(dLat / 2) +
        sin(dLon / 2) * sin(dLon / 2) * cos(lat1) * cos(lat2);
    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return R * c;
  }
}
